File: source/tests/python/throttle_tests.py
#!@PYTHON_EXECUTABLE@

import os
import time
import json
import unittest
import threading
import inspect
import numpy as np
import timemory as tim
from timemory import components as comp
from timemory.profiler import profile
from timemory.bundle import auto_timer, auto_tuple, marker

# --------------------------- test setup variables ----------------------------------- #


# --------------------------- helper functions ----------------------------------------- #

# check availability of a component
def check_available(component):
    return inspect.isclass(component)

# compute fibonacci
def fibonacci(n):
    return n if n < 2 else (fibonacci(n-1) + fibonacci(n-2))

# sleep for n nanosec
def do_sleep(n):
    time.sleep(n * 1e-9)

# cpu utilization for n nanosec
def consume(n):
    # consume cpu
    now = time.time_ns()
    while(time.time_ns() < (now + n)):
        pass

# get auto_tuple config
def get_config(items=["wall_clock"]):
    return [getattr(tim.component, x) for x in items]

# -------------------------- Thottle Tests set ---------------------------------------- #
# Throttle tests class
class TiMemoryThrottleTests(unittest.TestCase):
    # setup class: timemory settings
    @classmethod
    def setUpClass(self):
        # set up environment variables
        os.environ["TIMEMORY_VERBOSE"] = "1"
        os.environ["TIMEMORY_COLLAPSE_THREADS"] = "OFF"

        tim.settings.parse()
        
        tim.settings.verbose = 1
        tim.settings.debug = False
        tim.settings.json_output = True
        tim.settings.mpi_thread  = False
        tim.settings.file_output = False
        tim.settings.dart_output = True
        tim.settings.dart_count = 1
        tim.settings.banner = False
        
        tim.timemory_trace_init("wall_clock", False, "throttle_tests")

        self.nthreads = 4

    # Tear down class: timemory_finalize
    @classmethod
    def tearDownClass(self):
        # unset environment variables
        del os.environ['TIMEMORY_VERBOSE']
        del os.environ['TIMEMORY_COLLAPSE_THREADS']

        # tim.timemory_trace_finalize()
        # timemory finalize
        #tim.timemory_finalize()
        # tim.dmp.finalize()
        pass

# -------------------------------------------------------------------------------------- #
    # test expect_true
    def test_expect_true(self):
        """
        expect_true
        """
        n = 2 * tim.settings.throttle_count
        tim.timemory_push_trace("true")
        for i in range(n):
            tim.timemory_push_trace(self.shortDescription())
            tim.timemory_pop_trace(self.shortDescription())
        tim.timemory_pop_trace("true")

        self.assertTrue(tim.timemory_is_throttled(self.shortDescription()))

# -------------------------------------------------------------------------------------- #
    # test expect_false
    def test_expect_false(self):
        """
        expect_false
        """
        n = 2 * tim.settings.throttle_count
        v = 2 * tim.settings.throttle_value

        for i in range(n):
            tim.timemory_push_trace(self.shortDescription())
            consume(v)
            tim.timemory_pop_trace(self.shortDescription())

        self.assertFalse(tim.timemory_is_throttled(self.shortDescription()))

# -------------------------------------------------------------------------------------- #
    def test_region_serial(self):
        """
        region_serial
        """
        def _run(name):
            tim.timemory_push_region("rsthread")
            n = 8 * tim.settings.throttle_count
            for i in range(n):
                tim.timemory_push_region(name)
                tim.timemory_pop_region(name)

            tim.timemory_pop_region("rsthread")

        for i in range(self.nthreads):
            _run(self.shortDescription())
        
        #print(tim.timemory_is_throttled(self.shortDescription()))
        #print(tim.timemory_is_throttled("thread"))

# -------------------------------------------------------------------------------------- #
    # test region_multithreaded
    def test_region_multithreaded(self):
        """
        region_multithreaded
        """
        def _run(name):
            tim.timemory_push_region("rthread")
            n = 8 * tim.settings.throttle_count
            for i in range(n):
                tim.timemory_push_region(name)
                tim.timemory_pop_region(name)

            tim.timemory_pop_region("rthread")

        threads = []

        for i in range (self.nthreads):
            thd = threading.Thread(target=_run, args=(self.shortDescription(),))
            thd.start()
            threads.append(thd)

        for itr in threads:
            itr.join()

# -------------------------------------------------------------------------------------- #
    # test multithreaded
    def test_multithreaded(self):
        """
        multithreaded
        """
        # using tuple_t = tim::auto_tuple<tim::component::wall_clock>;

        # np.array of False
        is_throttled = np.full(self.nthreads, False)

        # _run function
        def _run(name, idx):
            tim.timemory_push_trace("mthread")
            n = 2 * tim.settings.throttle_count
            v = 2 * tim.settings.throttle_value
            if (idx % 2 == 1):
                for i in range(n):
                    tim.timemory_push_trace(name)
                    consume(v)
                    tim.timemory_pop_trace(name)
            else:
                for i in range(n):
                    tim.timemory_push_trace(name)
                    tim.timemory_pop_trace(name)

            tim.timemory_pop_trace("mthread")
            is_throttled[idx] = tim.timemory_is_throttled(name)

        # thread handles
        threads = []

        # make new threads
        for i in range (self.nthreads):
            thd = threading.Thread(target=_run, args=(self.shortDescription(), i))
            thd.start()
            threads.append(thd)

        # wait for join
        for itr in threads:
            itr.join()

        # check assertion
        for i in range (self.nthreads):
            _answer = False if (i % 2 == 1) else True
            print("thread " + str(i) + " throttling: " + str(is_throttled[i]))
            self.assertTrue(is_throttled[i] == _answer)

# -------------------------------------------------------------------------------------- #
    # test tuple_serial
    def test_tuple_serial(self):
        """
        tuple_serial
        """
        @marker(components=("wall_clock"), key="thread")
        def _run(name):
            with auto_tuple(get_config(["wall_clock"]), key="auto_tuple_serial"):
                n = 8 * tim.settings.throttle_count
                for i in range(n):
                    with marker(components=("wall_clock"), key=self.shortDescription()):
                        pass
            self.assertFalse(tim.timemory_is_throttled("thread"))
            self.assertFalse(tim.timemory_is_throttled(self.shortDescription()))

        # run with auto tuple (wall_clock)
        for i in range(self.nthreads):
            _run(self.shortDescription())

# -------------------------------------------------------------------------------------- #
    # test tuple_multithreaded
    def test_tuple_multithreaded(self):
        """
        tuple_multithreaded
        """
        @marker(components=("wall_clock"), key="thread")
        def _run(name):
            with auto_tuple(get_config(["wall_clock"]), key="auto_tuple_multithreaded"):
                n = 8 * tim.settings.throttle_count
                for i in range(n):
                    with marker(components=("wall_clock"), key=self.shortDescription()):
                        pass

            self.assertFalse(tim.timemory_is_throttled(self.shortDescription()))
            self.assertFalse(tim.timemory_is_throttled("thread"))

        # thread handles
        threads = []

        # make new threads
        for i in range (self.nthreads):
            thd = threading.Thread(target=_run, args=(self.shortDescription(), ))
            thd.start()
            threads.append(thd)

        # wait for join
        for itr in threads:
            itr.join()

# ----------------------------- main test runner ---------------------------------------- #
# main runner
if __name__ == '__main__':
    # run all tests
    unittest.main()

File: source/tests/throttle_tests.cpp
// MIT License
//
// Copyright (c) 2020, The Regents of the University of California,
// through Lawrence Berkeley National Laboratory (subject to receipt of any
// required approvals from the U.S. Dept. of Energy).  All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "test_macros.hpp"

TIMEMORY_TEST_DEFAULT_MAIN

#include "gtest/gtest.h"

#include <chrono>
#include <condition_variable>
#include <iostream>
#include <mutex>
#include <random>
#include <thread>
#include <vector>

#include "timemory/library.h"
#include "timemory/timemory.hpp"

//--------------------------------------------------------------------------------------//

namespace details
{
//--------------------------------------------------------------------------------------//
//  Get the current tests name
//
inline std::string
get_test_name()
{
    return std::string(::testing::UnitTest::GetInstance()->current_test_suite()->name()) +
           "." + ::testing::UnitTest::GetInstance()->current_test_info()->name();
}

// this function consumes approximately "n" milliseconds of real time
inline void
do_sleep(long n)
{
    std::this_thread::sleep_for(std::chrono::nanoseconds(n));
}

inline void
consume(long n)
{
    auto now = std::chrono::steady_clock::now();
    // try until time point
    while(std::chrono::steady_clock::now() < (now + std::chrono::nanoseconds(n)))
    {
    }
}
}  // namespace details

//--------------------------------------------------------------------------------------//

class throttle_tests : public ::testing::Test
{
    using wc_t = tim::component::wall_clock;

protected:
    static void SetUpTestSuite()
    {
        puts("[SetupTestSuite] setup starting");
        tim::set_env("TIMEMORY_VERBOSE", "1", 1);
        tim::set_env("TIMEMORY_COLLAPSE_THREADS", "OFF", 0);
        tim::settings::debug()       = false;
        tim::settings::json_output() = true;
        tim::settings::mpi_thread()  = false;
        puts("[SetupTestSuite] initializing dmp");
        tim::dmp::initialize(_argc, _argv);
        puts("[SetupTestSuite] initializing timemory");
        tim::timemory_init(_argc, _argv);
        tim::settings::dart_output() = true;
        tim::settings::dart_count()  = 1;
        tim::settings::banner()      = false;
        timemory_trace_init("wall_clock", false, "throttle_tests");
        puts("[SetupTestSuite] timemory initialized");
        tim::settings::verbose() = 1;
        tim::settings::debug()   = false;
        puts("[SetupTestSuite] setup completed");
        metric().start();
    }

    // always disable debug
    void SetUp() override
    {
        tim::settings::debug() = false;
        wc_init                = tim::storage<wc_t>::instance()->get();
        if(!wc_init.empty())
            offset = 0;
    }

    static void TearDownTestSuite()
    {
        metric().stop();
        print_dart(metric());
        timemory_trace_finalize();
        tim::timemory_finalize();
        if(tim::dmp::rank() == 0)
            tim::enable_signal_detection(tim::signal_settings::get_default());
        tim::dmp::finalize();
    }

    auto get_size_delta()
    {
        auto wc_data = tim::storage<wc_t>::instance()->get();
        return (wc_data.size() - wc_init.size());
    }

    auto get_count_delta(const std::string& _id)
    {
        auto _get_count = [&_id](auto&& _data) {
            int64_t _n = 0;
            for(auto&& itr : _data)
            {
                auto _lbl = tim::operation::decode<TIMEMORY_API>{}(
                    ::tim::get_hash_identifier(itr.hash()));
                if(_lbl == _id)
                {
                    _n += itr.data().get_laps();
                }
            }
            return _n;
        };

        return (_get_count(tim::storage<wc_t>::instance()->get()) - _get_count(wc_init));
    }

    std::string write_data()
    {
        if(tim::dmp::rank() > 0)
            return std::string{};

        std::ostringstream _oss;
        auto               _write_data = [&_oss](const auto& _data) {
            for(size_t i = 0; i < _data.size(); ++i)
                _oss << "    " << i << "/" << _data.size()
                     << " :: " << as_string(_data.at(i)) << "\n";
        };

        _oss << "Initial data:\n";
        _write_data(wc_init);
        _oss << "Current data:\n";
        _write_data(tim::storage<wc_t>::instance()->get());

        return _oss.str();
    }

    static constexpr uint64_t            nthreads = 4;
    size_t                               offset   = 1;
    std::vector<tim::node::result<wc_t>> wc_init{};
};

//--------------------------------------------------------------------------------------//

TEST_F(throttle_tests, expect_true)
{
    auto name = details::get_test_name();
    auto n    = 2 * tim::settings::throttle_count();

    for(size_t i = 0; i < n; ++i)
    {
        timemory_push_trace(name.c_str());
        timemory_pop_trace(name.c_str());
    }

    std::cout << std::boolalpha << "is_throttled(" << name
              << ") == " << timemory_is_throttled(name.c_str()) << std::endl;
    EXPECT_EQ(get_count_delta(name), tim::settings::throttle_count()) << write_data();
#if !defined(TIMEMORY_RELAXED_TESTING)
    EXPECT_TRUE(timemory_is_throttled(name.c_str())) << write_data();
#endif
}

//--------------------------------------------------------------------------------------//

TEST_F(throttle_tests, expect_false)
{
    auto name = details::get_test_name();
    auto n    = 2 * tim::settings::throttle_count();
    auto v    = 2 * tim::settings::throttle_value();

    for(size_t i = 0; i < n; ++i)
    {
        timemory_push_trace(name.c_str());
        // details::do_sleep(v);
        details::consume(v);
        timemory_pop_trace(name.c_str());
    }

    std::cout << std::boolalpha << "is_throttled(" << name
              << ") == " << timemory_is_throttled(name.c_str()) << std::endl;
    EXPECT_EQ(get_count_delta(name), 2 * tim::settings::throttle_count()) << write_data();
#if !defined(TIMEMORY_RELAXED_TESTING)
    EXPECT_FALSE(timemory_is_throttled(name.c_str())) << write_data();
#endif
}

//--------------------------------------------------------------------------------------//

TEST_F(throttle_tests, multithreaded)
{
    std::array<bool, nthreads> is_throttled;
    is_throttled.fill(false);

    auto _run = [&is_throttled](uint64_t idx) {
        timemory_push_trace("thread");
        auto name = details::get_test_name();
        auto n    = 2 * tim::settings::throttle_count();
        auto v    = 2 * tim::settings::throttle_value();
        if(idx % 2 == 1)
        {
            for(size_t i = 0; i < n; ++i)
            {
                timemory_push_trace(name.c_str());
                // details::do_sleep(v);
                details::consume(v);
                timemory_pop_trace(name.c_str());
            }
        }
        else
        {
            for(size_t i = 0; i < n; ++i)
            {
                timemory_push_trace(name.c_str());
                timemory_pop_trace(name.c_str());
            }
        }
        timemory_pop_trace("thread");
        is_throttled.at(idx) = timemory_is_throttled(name.c_str());
    };

    std::vector<std::thread> threads;
    for(uint64_t i = 0; i < nthreads; ++i)
        threads.emplace_back(_run, i);
    for(auto& itr : threads)
        itr.join();

    for(uint64_t i = 0; i < nthreads; ++i)
    {
        bool _answer = (i % 2 == 1) ? false : true;
        std::cout << "thread " << i << " throttling: " << std::boolalpha
                  << is_throttled[i] << ". expected: " << _answer << "\n";
        EXPECT_EQ(get_size_delta(), 2 * nthreads) << write_data();
#if !defined(TIMEMORY_RELAXED_TESTING)
        EXPECT_TRUE(is_throttled[i] == _answer) << write_data();
#endif
    }
}

//--------------------------------------------------------------------------------------//

TEST_F(throttle_tests, do_nothing)
{
    auto n = tim::settings::throttle_count();
    for(size_t i = 0; i < n; ++i)
        details::do_sleep(10);
    EXPECT_EQ(get_size_delta(), 0) << write_data();
}

//--------------------------------------------------------------------------------------//

TEST_F(throttle_tests, region_serial)
{
    auto _run = []() {
        timemory_push_region("thread");
        auto name = details::get_test_name();
        auto n    = 8 * tim::settings::throttle_count();
        for(size_t i = 0; i < n; ++i)
        {
            timemory_push_region(name.c_str());
            timemory_pop_region(name.c_str());
        }
        timemory_pop_region("thread");
    };

    for(uint64_t i = 0; i < nthreads; ++i)
        _run();

    EXPECT_EQ(get_size_delta(), 1 + offset) << write_data();
}

//--------------------------------------------------------------------------------------//

TEST_F(throttle_tests, region_multithreaded)
{
    auto _run = []() {
        timemory_push_region("thread");
        auto name = details::get_test_name();
        auto n    = 8 * tim::settings::throttle_count();
        for(size_t i = 0; i < n; ++i)
        {
            timemory_push_region(name.c_str());
            timemory_pop_region(name.c_str());
        }
        timemory_pop_region("thread");
    };

    std::vector<std::thread> threads;
    for(uint64_t i = 0; i < nthreads; ++i)
        threads.emplace_back(_run);
    for(auto& itr : threads)
        itr.join();
    EXPECT_EQ(get_size_delta(), 2 * nthreads) << write_data();
}

//--------------------------------------------------------------------------------------//

TEST_F(throttle_tests, tuple_serial)
{
    using tuple_t = tim::auto_tuple<tim::component::wall_clock>;
    auto _run     = []() {
        TIMEMORY_BLANK_MARKER(tuple_t, "thread");
        auto name = details::get_test_name();
        auto n    = 8 * tim::settings::throttle_count();
        for(size_t i = 0; i < n; ++i)
        {
            TIMEMORY_BLANK_MARKER(tuple_t, name);
        }
    };

    for(uint64_t i = 0; i < nthreads; ++i)
        _run();
    EXPECT_EQ(get_size_delta(), 1 + offset) << write_data();
}

//--------------------------------------------------------------------------------------//

TEST_F(throttle_tests, tuple_multithreaded)
{
    using tuple_t = tim::auto_tuple<tim::component::wall_clock>;
    auto _run     = []() {
        TIMEMORY_BLANK_MARKER(tuple_t, "thread");
        auto name = details::get_test_name();
        auto n    = 8 * tim::settings::throttle_count();
        for(size_t i = 0; i < n; ++i)
        {
            TIMEMORY_BLANK_MARKER(tuple_t, name);
        }
    };

    std::vector<std::thread> threads;
    for(uint64_t i = 0; i < nthreads; ++i)
        threads.emplace_back(_run);
    for(auto& itr : threads)
        itr.join();
    EXPECT_EQ(get_size_delta(), 2 * nthreads) << write_data();
}

//--------------------------------------------------------------------------------------//
